### PR TITLE
Taxation calculation fix

### PR DESCRIPTION
**Pull request by GitGubM8**
# Depends on:
- https://github.com/Thrillberg/imperial/pull/688
[MinDiff](https://github.com/GitGubM8/imperial/compare/TaxStatus...GitGubM8:imperial:Taxation_Fix)

## Source:
- In Imperial, your nation cant ever lose money
image

- In Imperial 2030 however, you pay out of treasury until done or bankrupt
image
How I interpreted it is that, even if your taxation revenue breakeven with the soldiers' pay, the country has to pay the player the bonus from its treasury
image

## Changes:
- In Imperial 2030, your nation can now lose money from taxation if you really want to (helps avoid Swiss Banker hounding you)

## Screenshots:
- Country losing money:
image
image
image
- Country going bankrupt:
image
image
image
- Taxation breakeven with soldier pay, country paying player bonus from treasury:
image
image
image

### CODE REVIEW

**Comment by GitGubM8** [comment]
One thing I didn't test was if we needed to migrate older/in-progress games, although from my understanding, I don't think they'd break

Also not sure if I needed to implement the `undo` somewhere

File: app/javascript/lib/imperial.js [hunk 1]
```diff
@@ -1277,110 +1277,36 @@ export default class Imperial {
...
-        }
-        // Player receives full excess taxes
-        this.players[this.currentPlayerName].cash += excessTaxes;
+        const bonus = this.playerBonusAfterUnitMaintenanceCosts(nationName, taxes);
+
+        this.players[this.currentPlayerName].cash += bonus;
         this.annotatedLog.push(
           Action.playerGainsCash({
             player: this.currentPlayerName,
-            amount: excessTaxes,
```
Comment (GitGubM8) [inline app/javascript/lib/imperial.js:1295]: Also fixes the double logging of tax bonuses for `Imperial 2030` & `Imperial Asia`, the second log being https://github.com/Thrillberg/imperial/pull/689/files#diff-005e3c79c3c0db9f418406044629ce50e2d95e46a13ce76da9663ec5b37ad77fL1329
image

**Comment by GitGubM8** [comment]
[8224097](https://github.com/Thrillberg/imperial/pull/689/commits/822409780a9fbd55ae2ccb7d713d925128620e9e)
image
image

**Comment by Thrillberg** [comment]
@GitGubM8 I think I'll wait to review this and the other PR until the first one is merged and we do a rebase so that these are smaller. Does that sound ok? In the future, I think it's easier if we have each PR be self-contained (ie, branching off `main`), rather than build on another open PR. Also, I gave you access to the main repo so you can just branch and PR off that, rather than your fork, if you prefer! 😄

**Comment by GitGubM8** [comment]
> @GitGubM8 I think I'll wait to review this and the other PR until the first one is merged and we do a rebase so that these are smaller. Does that sound ok? In the future, I think it's easier if we have each PR be self-contained (ie, branching off `main`), rather than build on another open PR. Also, I gave you access to the main repo so you can just branch and PR off that, rather than your fork, if you prefer! 😄

Sounds good, there's also a MinDiff at the top of it helps

File: app/javascript/lib/imperial.js [hunk 2]
```diff
@@ -1277,110 +1277,36 @@ export default class Imperial {
...
         nation.taxChartPosition = this.getTaxChartPosition(taxes);
-        let excessTaxes = nation.taxChartPosition - previousTaxChartPosition;
-        // Players can never lose money here
-        if (excessTaxes < 0) {
-          excessTaxes = 0;
-        }
-        // Player receives full excess taxes
-        this.players[this.currentPlayerName].cash += excessTaxes;
+        nation.treasury += nationProfit;
+        // can be less than 0m
```
Comment (Thrillberg) [inline app/javascript/lib/imperial.js:1290]: Really? Nations can lose money?

File: app/javascript/lib/imperial2030.test.js [hunk 3]
```diff
@@ -143,7 +143,7 @@ describe('imperial2030', () => {
         expect(game.nations.get(Nation2030.RU).treasury).toEqual(9);
       });
 
-      test('nations cannot be paid less than 0 if they control many units', () => {
+      test('nations lose money if they control many units', () => {
         const game = newGame();
         // Arbitrarily give RU 5 treasury; we want this to increase by 1
```
Comment (Thrillberg) [inline app/javascript/lib/imperial2030.test.js]: If this comment is incorrect, please fix it.

EDIT: just to clarify I mean line 148

File: app/javascript/lib/imperial2030.test.js [hunk 4]
```diff
@@ -143,7 +143,7 @@ describe('imperial2030', () => {
         expect(game.nations.get(Nation2030.RU).treasury).toEqual(9);
       });
 
-      test('nations cannot be paid less than 0 if they control many units', () => {
+      test('nations lose money if they control many units', () => {
```
Comment (Thrillberg) [inline app/javascript/lib/imperial2030.test.js:146]: Can nations lose money here? Does the last sentence from the rules here contradict this?
Screenshot 2023-03-19 at 7 03 08 PM

File: app/javascript/src/components/GameLogEntry.vue [hunk 5]
```diff
@@ -165,7 +165,7 @@ export default {
         case 'playerGainsCash':
           return `${action.payload.player} gained ${action.payload.amount}m in taxes.`;
         case 'nationGainsTreasury':
-          return `${stringify(action.payload.nation.value)} gained ${action.payload.amount}m in taxes.`;
+          return `${stringify(action.payload.nation.value)} has changed by ${action.payload.amount}m in taxes.`;
```
Comment (Thrillberg) [inline app/javascript/src/components/GameLogEntry.vue:169]: If nations cannot lose money, then this change shouldn't be made. Otherwise it should remain.

**Review by Thrillberg — COMMENTED** [review]
Just one question about whether nations can lose money. Otherwise, looks great to me!

EDIT! Ah sorry I just saw your comments above about the rules. I'll check them out. I'm sure you're right!!

**Review by Thrillberg — APPROVED** [review]
Ok on another look, this is fine and no changes needed! Thank you!!

**Comment by Thrillberg** [comment]
If we merge this, will the prior PR become redundant (ie, we should delete it)?

File: app/javascript/lib/imperial2030.test.js (hunk 3 above)
Comment (GitGubM8) [reply app/javascript/lib/imperial2030.test.js]: Good catch didn't notice it, will update

Comment (GitGubM8) [reply app/javascript/lib/imperial2030.test.js]: [67de8dc](https://github.com/Thrillberg/imperial/pull/689/commits/67de8dc92e10c3a6a3c9ecdd3638707aa7953619)

**Comment by GitGubM8** [comment]
> If we merge this, will the prior PR become redundant (ie, we should delete it)?

I merged the prior one first, so we get one last opportunity to review this one with a lighter diff, you can do the honors whenever you're ready

**Comment by GitGubM8** [comment]
> One thing I didn't test was if we needed to migrate older/in-progress games, although from my understanding, I don't think they'd break
> 
> Also not sure if I needed to implement the `undo` somewhere

Also wanted to bring attention to this, just in case

**Comment by Thrillberg** [comment]
>One thing I didn't test was if we needed to migrate older/in-progress games, although from my understanding, I don't think they'd break

@GitGubM8 Older games _may_ break but that's ok. I don't go to any extra effort to migrate in-progress games to such changes and just hope people are understanding about the game adjusting itself to the new rules. So far they have been! This usually results in a game "rewinding" to some point before certain moves became illegal. In our case, I doubt even that will happen, unless this rule got exercised (which would indicate that the game should be rewound anyway).

Thanks for thinking of this concern though!!

>Also not sure if I needed to implement the `undo` somewhere

Nope, all `undo` does is pretend like the "undone" move was never made in the first place. So we don't need to be thinking about "reverse-logic" or anything like that! 😄

**Pull request merged**